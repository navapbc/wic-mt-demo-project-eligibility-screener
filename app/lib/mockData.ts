import cloneDeep from 'lodash/cloneDeep'

import type { SessionData } from '@src/types'
import { initialSessionData } from '@utils/sessionData'

// Helper functions to create valid data for each section of the session.
export function getMockEligibilityData() {
  return {
    residential: 'yes',
    categorical: ['pregnant', 'guardian'],
    previouslyEnrolled: 'no',
    adjunctive: ['fdpir', 'snap'],
  }
}

export function getMockChooseClinicData() {
  return {
    zipCode: '12345',
    clinic: {
      agency: 'AGENCY ZERO',
      agencyAddress: '0000 St, Helena, MT 00000',
      agencyTelephone: '(000) 000-0000',
      clinic: 'CLINIC ZERO',
      clinicAddress: '0000 St, Helena, MT 00000',
      clinicTelephone: '(000) 000-0000',
      county: 'COUNTY ZERO',
      id: 0,
      zip: '00000',
    },
  }
}

export function getMockIncomeData() {
  return {
    householdSize: '3',
  }
}

export function getMockContactData() {
  return {
    firstName: 'Jack',
    lastName: 'O Lantern',
    phone: '1231231234',
    comments: 'comments',
  }
}

// Helper function to setup a fully filled out session.
export function getMockSessionData() {
  const mockSession = getEmptyMockSession()
  mockSession.eligibility = getMockEligibilityData()
  mockSession.chooseClinic = getMockChooseClinicData()
  mockSession.contact = getMockContactData()
  return mockSession
}

// Helper function te setup an empty session that passes typechecking.
export function getEmptyMockSession(): SessionData {
  return cloneDeep(initialSessionData)
}
