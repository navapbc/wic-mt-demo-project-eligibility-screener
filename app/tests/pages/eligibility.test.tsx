import { render, screen } from '@testing-library/react'
import singletonRouter from 'next/router'

import Eligibility from '@pages/eligibility'

import { getMockEligibilityData } from '../helpers/mockData'
import { setMockSession, setup } from '../helpers/setup'
import {
  testAccessibility,
  testActionButtonReviewMode,
  testActionButtonRoute,
  testBackLink,
  testSnapshot,
} from '../helpers/sharedTests'
import { invalidEligibilityCombinations } from '../utils/dataValidation/isValidEligibility.test'

/**
 * Test setup
 */

const route = '/eligibility'
const backRoute = '/how-it-works'
const forwardRoute = '/income'

/**
 * Begin tests
 */

it('should match full page snapshot', () => {
  const { mockSession } = setup(route)
  testSnapshot(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )
})

it('should pass accessibility scan', async () => {
  const { mockSession } = setup(route)
  await testAccessibility(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )
})

it('action button should render differently in review mode', () => {
  const { mockSession } = setup(route)
  testActionButtonReviewMode(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      reviewMode={true}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />,
    route
  )
})

it('should have a back link that matches the backRoute in default mode', () => {
  const { mockSession } = setup(route)
  testBackLink(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />,
    backRoute
  )
})

it('should have a back link that matches the backRoute in review mode', () => {
  const { mockSession } = setup(route)
  testBackLink(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      reviewMode={true}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />,
    backRoute
  )
})

it('should have an action button that routes to forwardRoute', async () => {
  const { mockSession, user } = setup(route)
  mockSession.eligibility = getMockEligibilityData()
  const element = (
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )
  await testActionButtonRoute(element, forwardRoute, 'Continue', user)
})

it.each(invalidEligibilityCombinations)(
  'action button should be disabled if %s',
  (description, residential, categorical, previouslyEnrolled, adjunctive) => {
    const { mockSession } = setup(route)

    // Set the initial mock session values
    mockSession.eligibility.residential = residential as string
    mockSession.eligibility.categorical = categorical as string[]
    mockSession.eligibility.previouslyEnrolled = previouslyEnrolled as string
    mockSession.eligibility.adjunctive = adjunctive as string[]
    render(
      <Eligibility
        session={mockSession}
        setSession={setMockSession}
        backRoute={backRoute}
        forwardRoute={forwardRoute}
      />
    )

    // Check the button is disabled.
    const button = screen.getByRole('button', { name: /Continue/i })
    expect(button).toBeDisabled()
  }
)

it('action button should be enabled if all requirements are met', () => {
  const { mockSession } = setup(route)

  // Set the initial mock session values
  mockSession.eligibility.residential = 'anything'
  mockSession.eligibility.categorical = ['anything']
  mockSession.eligibility.previouslyEnrolled = 'anything'
  mockSession.eligibility.adjunctive = ['anything']
  render(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )

  // Check the button is enabled.
  const button = screen.getByRole('button', { name: /Continue/i })
  expect(button).not.toBeDisabled()
})

it('should display user values on refresh/page load', () => {
  const { mockSession } = setup(route)
  mockSession.eligibility.residential = 'yes'
  mockSession.eligibility.categorical = ['pregnant']
  mockSession.eligibility.previouslyEnrolled = 'yes'
  mockSession.eligibility.adjunctive = ['tanf']
  render(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )

  const residential = screen.getAllByRole('radio', { name: /Yes/i })[0]
  expect(residential).toBeChecked()

  const categorical = screen.getByRole('checkbox', {
    name: /I'm pregnant/i,
  })
  expect(categorical).toBeChecked()

  const previouslyEnrolled = screen.getAllByRole('radio', {
    name: /Yes/i,
  })[1]
  expect(previouslyEnrolled).toBeChecked()

  const adjunctive = screen.getByRole('checkbox', { name: /tanf/i })
  expect(adjunctive).toBeChecked()
})

it('action button should stay disabled until all requirements are met and re-disable if reqirements are unmet', async () => {
  const { mockSession, user } = setup(route)

  render(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )
  const button = screen.getByRole('button', { name: /Continue/i })
  expect(button).toBeDisabled()

  // Select a residental option.
  const residential = screen.getAllByRole('radio', { name: /Yes/i })[0]
  expect(residential).not.toBeChecked()
  await user.click(residential)
  expect(residential).toBeChecked()
  expect(button).toBeDisabled()

  // Select a categorical option.
  const categorical = screen.getByRole('checkbox', {
    name: /I'm pregnant/i,
  })
  expect(categorical).not.toBeChecked()
  await user.click(categorical)
  expect(categorical).toBeChecked()
  expect(button).toBeDisabled()

  // Select a previouslyEnrolled option.
  const previouslyEnrolled = screen.getAllByRole('radio', {
    name: /Yes/i,
  })[1]
  expect(previouslyEnrolled).not.toBeChecked()
  await user.click(previouslyEnrolled)
  expect(previouslyEnrolled).toBeChecked()
  expect(button).toBeDisabled()

  // Select a adjunctive option.
  const adjunctive = screen.getByRole('checkbox', { name: /FDPIR/i })
  expect(adjunctive).not.toBeChecked()
  await user.click(adjunctive)
  expect(adjunctive).toBeChecked()
  // The button should finally be enabled.
  expect(button).not.toBeDisabled()

  // De-select a categorical option.
  await user.click(categorical)
  expect(categorical).not.toBeChecked()
  // The button should not be disabled again.
  expect(button).toBeDisabled()

  // Re-select a categorical option.
  await user.click(categorical)
  expect(categorical).toBeChecked()
  expect(button).not.toBeDisabled()

  // De-select a adjunctive option.
  await user.click(adjunctive)
  expect(adjunctive).not.toBeChecked()
  // The button should not be disabled again.
  expect(button).toBeDisabled()

  // Re-select a adjunctive option.
  await user.click(adjunctive)
  expect(adjunctive).toBeChecked()
  expect(button).not.toBeDisabled()
})

it('should uncheck all other checkboxes when None of the Above is checked', async () => {
  const { mockSession, user } = setup(route)
  render(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )

  const fdpir = screen.getByRole('checkbox', { name: /FDPIR/i })
  await user.click(fdpir)
  const none = screen.getAllByRole('checkbox', { name: /None/i })[1]
  await user.click(none)

  expect(fdpir).not.toBeChecked()
  expect(none).toBeChecked()
})

it('should uncheck None of the Above if another option is checked', async () => {
  const { mockSession, user } = setup(route)
  render(
    <Eligibility
      session={mockSession}
      setSession={setMockSession}
      backRoute={backRoute}
      forwardRoute={forwardRoute}
    />
  )

  const none = screen.getAllByRole('checkbox', { name: /None/i })[1]
  await user.click(none)
  const fdpir = screen.getByRole('checkbox', { name: /FDPIR/i })
  await user.click(fdpir)

  expect(fdpir).toBeChecked()
  expect(none).not.toBeChecked()
})
