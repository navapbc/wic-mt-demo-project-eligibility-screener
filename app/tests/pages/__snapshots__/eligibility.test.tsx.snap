// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match full page snapshot 1`] = `
<div>
  <a
    class="usa-link"
    href="/how-it-works"
  >
    Back
  </a>
  <h1>
    Check your eligibility
  </h1>
  <p>
    All required questions are marked with an asterisk
     (
    <abbr
      class="usa-hint usa-hint--required"
    >
      *
    </abbr>
    ).
  </p>
  <form
    class="usa-form usa-form--large"
  >
    <fieldset
      class="usa-fieldset"
    >
      <h2>
        Do you live or work in Montana?
        <abbr
          class="usa-hint usa-hint--required"
        >
           *
        </abbr>
      </h2>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="residential-yes"
          name="residential"
          type="radio"
          value="yes"
        />
        <label
          class="usa-radio__label"
          for="residential-yes"
        >
          Yes
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="residential-no"
          name="residential"
          type="radio"
          value="no"
        />
        <label
          class="usa-radio__label"
          for="residential-no"
        >
          No
        </label>
      </div>
    </fieldset>
    <fieldset
      class="usa-fieldset"
    >
      <h2>
        Please select all the following that apply to your household:
        <abbr
          class="usa-hint usa-hint--required"
        >
           *
        </abbr>
      </h2>
      <div
        class="usa-accordion usa-accordion--bordered"
      >
        <h3
          class="usa-accordion__heading"
        >
          <button
            aria-controls="b-a1"
            aria-expanded="false"
            class="usa-accordion__button"
            type="button"
          >
            Why are we asking this?
          </button>
        </h3>
        <div
          class="usa-accordion__content"
          hidden=""
          id="b-a1"
        >
          There are multiple categories of eligibility, and your answers here help us understand what benefits you and/or your household may qualify for.
        </div>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="categorical-pregnant"
          name="categorical"
          type="checkbox"
          value="pregnant"
        />
        <label
          class="usa-checkbox__label"
          for="categorical-pregnant"
        >
          I'm pregnant
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="categorical-baby"
          name="categorical"
          type="checkbox"
          value="baby"
        />
        <label
          class="usa-checkbox__label"
          for="categorical-baby"
        >
          I've had a baby in the past 12 months
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="categorical-child"
          name="categorical"
          type="checkbox"
          value="child"
        />
        <label
          class="usa-checkbox__label"
          for="categorical-child"
        >
          I'm the parent of an infant or child under the age of 5 years old
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="categorical-guardian"
          name="categorical"
          type="checkbox"
          value="guardian"
        />
        <label
          class="usa-checkbox__label"
          for="categorical-guardian"
        >
          I'm the guardian or foster parent of an infant or child under the age of 5 years old
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="categorical-loss"
          name="categorical"
          type="checkbox"
          value="loss"
        />
        <label
          class="usa-checkbox__label"
          for="categorical-loss"
        >
          I have had a pregnancy end in the last 6 months
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="categorical-none"
          name="categorical"
          type="checkbox"
          value="none"
        />
        <label
          class="usa-checkbox__label"
          for="categorical-none"
        >
          None of the above
        </label>
      </div>
    </fieldset>
    <fieldset
      class="usa-fieldset"
    >
      <h2>
        Have you or someone in your household received WIC benefits before?
        <abbr
          class="usa-hint usa-hint--required"
        >
           *
        </abbr>
      </h2>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="previouslyEnrolled-yes"
          name="previouslyEnrolled"
          type="radio"
          value="yes"
        />
        <label
          class="usa-radio__label"
          for="previouslyEnrolled-yes"
        >
          Yes
        </label>
      </div>
      <div
        class="usa-radio"
      >
        <input
          class="usa-radio__input usa-radio__input--tile"
          id="previouslyEnrolled-no"
          name="previouslyEnrolled"
          type="radio"
          value="no"
        />
        <label
          class="usa-radio__label"
          for="previouslyEnrolled-no"
        >
          No
        </label>
      </div>
    </fieldset>
    <fieldset
      class="usa-fieldset"
    >
      <h2>
        Are you or someone in your household currently enrolled in any of the following programs in Montana?
        <abbr
          class="usa-hint usa-hint--required"
        >
           *
        </abbr>
      </h2>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="adjunctive-insurance"
          name="adjunctive"
          type="checkbox"
          value="insurance"
        />
        <label
          class="usa-checkbox__label"
          for="adjunctive-insurance"
        >
          Medicaid/Healthy Montana Kids Plus
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="adjunctive-snap"
          name="adjunctive"
          type="checkbox"
          value="snap"
        />
        <label
          class="usa-checkbox__label"
          for="adjunctive-snap"
        >
          SNAP (Supplemental Nutrition Assistance Program)
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="adjunctive-tanf"
          name="adjunctive"
          type="checkbox"
          value="tanf"
        />
        <label
          class="usa-checkbox__label"
          for="adjunctive-tanf"
        >
          TANF (Temporary Assistance for Needy Families)
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="adjunctive-fdpir"
          name="adjunctive"
          type="checkbox"
          value="fdpir"
        />
        <label
          class="usa-checkbox__label"
          for="adjunctive-fdpir"
        >
          FDPIR (Food Distribution Program on Indian Reservations)
        </label>
      </div>
      <div
        class="usa-checkbox"
      >
        <input
          class="usa-checkbox__input usa-checkbox__input--tile"
          id="adjunctive-none"
          name="adjunctive"
          type="checkbox"
          value="none"
        />
        <label
          class="usa-checkbox__label"
          for="adjunctive-none"
        >
          None of the above
        </label>
      </div>
    </fieldset>
    <button
      class="usa-button usa-button--small display-block margin-top-6"
      disabled=""
    >
      Continue
    </button>
  </form>
</div>
`;
